Zend/zend_objects.c

```diff
@@ -52,7 +52,18 @@ ZEND_API void zend_object_std_dtor(zend_object *object)
 	if (EXPECTED(object->ce->default_properties_count)) {
 		end = p + object->ce->default_properties_count;
 		do {
-			i_zval_ptr_dtor(p);
+			if (Z_REFCOUNTED_P(p)) {
+				if (UNEXPECTED(Z_ISREF_P(p))) {
+					zend_property_info *prop_info;
+					ZEND_REF_FOREACH_TYPE_SOURCES(Z_REF_P(p), prop_info) {
+						if (prop_info->ce == object->ce && p == OBJ_PROP(object, prop_info->offset)) {
```

Comment:
The CE check here is not right if the property is inherited. The following code generates a TypeError, even though it shouldn't.

```
class A {
    public int $prop = 42;
}
class B extends A {
}

$b = new B;
$r =& $b->prop;
unset($b);
$r = "foo";
```

Reply:
Another test-case involving shadowing:

```php
class A {
    private int $prop = 42;

    public function &getRef() {
        return $this->prop;
    }
}
class B extends A {
    public $prop;
}

$b = new B;
$r =& $b->getRef();
unset($b);
$r = "foo";
```

I think that just not checking the CE *should* be safe. Acquiring an OOB pointer is *technically* UB but it should not cause miscompiles here, as we're in struct hack land anyway.

Reply:
On a related note, I think it would be useful if `zend_reference_destroy()` checked if there are any sources left in debug mode and print a message. That should make it much easier to discover cases where we are not properly deleting references.

Reply:
`OBJ_PROP(object, prop_info->offset)` may simply point into unallocated memory (or that location be _accidentally_ equal to `p`), and then it _will_ blow up. Thus checking is necessary in any case.

So, looks like it's going to need an instanceof check

Reply:
Hm... you are right that it could accidentally equal `p`. I'm not sure if a simple instanceof will cut it though. In particular I'm thinking about the case where a property is redeclared in a child class, which IIRC shares the offset. So there might be to prop_infos with the same offset coming from two different classes where one is a child of the other. Not totally sure though.

Reply:
Yeah, just checked, properties which are protected in parent and public in child will generate two different properties with the same offset.

It actually is indistinguishable at this point whether the particular reference here has been created by the child or the parent.

Consider:
```php
class A {
    protected int $prop;
    public setRef(&$ref) {
        $this->prop = $ref;
    }
}
class B extends A {
    public int $prop;
}
$a = new A;
$b = new B;
$bb = new B;
$ref = 1;
$a->setRef($ref);
$b->setRef($ref);
$bb->prop = &$ref;
```
Now we have twice `A::$prop` and once `B::$prop` in the reference set. Without additional information, there is no way to know which one to remove when destroying $b or $bb...
The lacking piece of information would be the actual `zend_object *` pointer for non-static properties. Then we could easily compare the object pointers and we then know that it's safe to remove.

Reply:
Ah. Okay, at least we have that :-D

This is annoying though, we'll have to iterate over all the property infos ... And check whose ce is the closest to the object ce.

I'm not sure what the nicest way to implement that is...